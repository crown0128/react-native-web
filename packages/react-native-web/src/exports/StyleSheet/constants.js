export const STYLE_ELEMENT_ID = 'react-native-stylesheet';

export const STYLE_GROUPS = {
  reset: 0,
  modality: 0.1,
  classic: 1,
  atomic: 2.2,
  custom: {
    borderColor: 2,
    borderRadius: 2,
    borderStyle: 2,
    borderWidth: 2,
    display: 2,
    flex: 2,
    margin: 2,
    overflow: 2,
    overscrollBehavior: 2,
    padding: 2,

    marginHorizontal: 2.1,
    marginVertical: 2.1,
    paddingHorizontal: 2.1,
    paddingVertical: 2.1
  }
};
