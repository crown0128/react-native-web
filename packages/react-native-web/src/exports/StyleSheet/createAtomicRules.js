import createRuleBlock from './createRuleBlock';

const createAtomicRules = (selector, prop, value) => {
  const rules = [];

  switch (prop) {
    // pointerEvents is a special case that requires custom values and additional rules
    // See #513
    case 'pointerEvents': {
      let val = value;
      if (value === 'auto' || value === 'box-only') {
        val = 'auto !important';
        if (value === 'box-only') {
          const block = createRuleBlock({ [prop]: 'none' });
          rules.push(`${selector} > *{${block}}`);
        }
      } else if (value === 'none' || value === 'box-none') {
        val = 'none !important';
        if (value === 'box-none') {
          const block = createRuleBlock({ [prop]: 'auto' });
          rules.push(`${selector} > *{${block}}`);
        }
      }
      const block = createRuleBlock({ [prop]: val });
      rules.push(`${selector}{${block}}`);
      break;
    }

    case 'placeholderTextColor': {
      const block = createRuleBlock({ color: value, opacity: 1 });
      rules.push(`@media all {
${selector}::-webkit-input-placeholder{${block}}
${selector}::-moz-placeholder{${block}}
${selector}:-ms-input-placeholder{${block}}
${selector}::placeholder{${block}}
}`);
      break;
    }

    default: {
      const block = createRuleBlock({ [prop]: value });
      rules.push(`${selector}{${block}}`);
    }
  }

  return rules;
};

export default createAtomicRules;
