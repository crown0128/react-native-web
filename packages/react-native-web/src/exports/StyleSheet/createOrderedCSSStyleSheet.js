/**
 * Copyright (c) Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 */

type Groups = { [key: number]: Array<string> };
type Selectors = { [key: string]: boolean };

const slice = Array.prototype.slice;

/**
 * Order-based insertion of CSS.
 *
 * Each rule can be inserted (appended) into a numerically defined group.
 * Groups are ordered within the style sheet according to their number, with the
 * lowest first.
 *
 * Groups are implemented using Media Query blocks. CSSMediaRule implements the
 * CSSGroupingRule, which includes 'insertRule', allowing groups to be treated as
 * a sub-sheet.
 * https://developer.mozilla.org/en-US/docs/Web/API/CSSMediaRule
 *
 * The selector of the first rule of each group is used only to encode the group
 * number for hydration.
 */
export default function createOrderedCSSStyleSheet(sheet: ?CSSStyleSheet) {
  const groups: Groups = {};
  const selectors: Selectors = {};

  /**
   * Hydrate approximate record from any existing rules in the sheet.
   */
  if (sheet != null) {
    slice.call(sheet.cssRules).forEach(mediaRule => {
      if (mediaRule.media == null) {
        throw new Error(
          'OrderedCSSStyleSheet: hydrating invalid stylesheet. Expected only @media rules.'
        );
      }

      // Create group number
      const group = decodeGroupRule(mediaRule);
      groups[group] = [];

      // Create record of existing selectors and rules
      slice.call(mediaRule.cssRules).forEach(rule => {
        const selectorText = getSelectorText(rule.cssText);
        if (selectorText != null) {
          selectors[selectorText] = true;
          groups[group].push(rule.cssText);
        }
      });
    });
  }

  const OrderedCSSStyleSheet = {
    /**
     * The textContent of the style sheet.
     * Each group's rules are wrapped in a media query.
     */
    getTextContent(): string {
      return getOrderedGroups(groups)
        .map(group => {
          const rules = groups[group];
          const str = rules.join('\n');
          return createMediaRule(str);
        })
        .join('\n');
    },

    /**
     * Insert a rule into a media query in the style sheet
     */
    insert(cssText: string, group: number) {
      // Create a new group.
      if (groups[group] == null) {
        const markerRule = encodeGroupRule(group);

        // Create the internal record.
        groups[group] = [];
        groups[group].push(markerRule);

        // Create CSSOM CSSMediaRule.
        if (sheet != null) {
          const groupIndex = getOrderedGroups(groups).indexOf(group);
          insertRuleAt(sheet, createMediaRule(markerRule), groupIndex);
        }
      }

      // selectorText is more reliable than cssText for insertion checks. The
      // browser excludes vendor-prefixed properties and rewrites certain values
      // making cssText more likely to be different from what was inserted.
      const selectorText = getSelectorText(cssText);
      if (selectorText != null && selectors[selectorText] == null) {
        // Update the internal records.
        selectors[selectorText] = true;
        groups[group].push(cssText);
        // Update CSSOM CSSMediaRule.
        if (sheet != null) {
          const groupIndex = getOrderedGroups(groups).indexOf(group);
          const root = sheet.cssRules[groupIndex];
          if (root != null) {
            // $FlowFixMe: Flow is missing CSSOM types
            insertRuleAt(root, cssText, root.cssRules.length);
          }
        }
      }
    }
  };

  return OrderedCSSStyleSheet;
}

/**
 * Helper functions
 */

function createMediaRule(content) {
  return `@media all {\n${content}\n}`;
}

function encodeGroupRule(group) {
  return `[stylesheet-group="${group}"]{}`;
}

function decodeGroupRule(mediaRule) {
  return mediaRule.cssRules[0].selectorText.split('"')[1];
}

function getOrderedGroups(obj: { [key: number]: any }) {
  return Object.keys(obj)
    .sort()
    .map(k => Number(k));
}

const pattern = /\s*([,])\s*/g;
function getSelectorText(cssText) {
  const selector = cssText.split('{')[0].trim();
  return selector !== '' ? selector.replace(pattern, '$1') : null;
}

function insertRuleAt(root, cssText: string, position: number) {
  try {
    // $FlowFixMe: Flow is missing CSSOM types needed to type 'root'.
    root.insertRule(cssText, position);
  } catch (e) {
    // JSDOM doesn't support `CSSSMediaRule#insertRule`.
    // Also ignore errors that occur from attempting to insert vendor-prefixed selectors.
  }
}
