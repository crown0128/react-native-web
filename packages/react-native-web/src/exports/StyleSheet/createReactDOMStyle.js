/**
 * Copyright (c) Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import normalizeValueWithProperty from './normalizeValueWithProperty';

/**
 * The browser implements the CSS cascade, where the order of properties is a
 * factor in determining which styles to paint. React Native is different. It
 * gives giving precedence to the more specific style property. For example,
 * the value of `paddingTop` takes precedence over that of `padding`.
 *
 * This module creates mutally exclusive style declarations by expanding all of
 * React Native's supported shortform properties (e.g. `padding`) to their
 * longfrom equivalents.
 */

const emptyObject = {};
const styleShortFormProperties = {
  borderColor: ['borderTopColor', 'borderRightColor', 'borderBottomColor', 'borderLeftColor'],
  borderRadius: [
    'borderTopLeftRadius',
    'borderTopRightRadius',
    'borderBottomRightRadius',
    'borderBottomLeftRadius'
  ],
  borderStyle: ['borderTopStyle', 'borderRightStyle', 'borderBottomStyle', 'borderLeftStyle'],
  borderWidth: ['borderTopWidth', 'borderRightWidth', 'borderBottomWidth', 'borderLeftWidth'],
  margin: ['marginTop', 'marginRight', 'marginBottom', 'marginLeft'],
  marginHorizontal: ['marginRight', 'marginLeft'],
  marginVertical: ['marginTop', 'marginBottom'],
  overflow: ['overflowX', 'overflowY'],
  overscrollBehavior: ['overscrollBehaviorX', 'overscrollBehaviorY'],
  padding: ['paddingTop', 'paddingRight', 'paddingBottom', 'paddingLeft'],
  paddingHorizontal: ['paddingRight', 'paddingLeft'],
  paddingVertical: ['paddingTop', 'paddingBottom'],
  writingDirection: ['direction']
};

const borderWidthProps = {
  borderWidth: true,
  borderTopWidth: true,
  borderRightWidth: true,
  borderBottomWidth: true,
  borderLeftWidth: true
};

const monospaceFontStack = 'monospace, monospace';
const systemFontStack =
  'system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Ubuntu, "Helvetica Neue", sans-serif';

/**
 * Transform
 */

// { scale: 2 } => 'scale(2)'
// { translateX: 20 } => 'translateX(20px)'
const mapTransform = transform => {
  const type = Object.keys(transform)[0];
  const value = normalizeValueWithProperty(transform[type], type);
  return `${type}(${value})`;
};

// [1,2,3,4,5,6] => 'matrix3d(1,2,3,4,5,6)'
const convertTransformMatrix = transformMatrix => {
  const matrix = transformMatrix.join(',');
  return `matrix3d(${matrix})`;
};

const resolveTransform = (resolvedStyle, style) => {
  let transform = style.transform;
  if (Array.isArray(style.transform)) {
    transform = style.transform.map(mapTransform).join(' ');
  } else if (style.transformMatrix) {
    transform = convertTransformMatrix(style.transformMatrix);
  }
  resolvedStyle.transform = transform;
};

/**
 * Reducer
 */

const createReactDOMStyle = style => {
  if (!style) {
    return emptyObject;
  }

  const resolvedStyle = {};

  Object.keys(style)
    .sort()
    .forEach(prop => {
      let value = normalizeValueWithProperty(style[prop], prop);

      // Make sure the default border width is explicitly set to '0' to avoid
      // falling back to any unwanted user-agent styles.
      if (borderWidthProps[prop]) {
        value = value == null ? normalizeValueWithProperty(0) : value;
      }

      // Ignore everything else with a null value
      if (value == null) {
        return;
      }

      switch (prop) {
        // Ignore some React Native styles
        case 'aspectRatio':
        case 'elevation':
        case 'overlayColor':
        case 'resizeMode':
        case 'tintColor': {
          break;
        }

        // TODO: remove once this issue is fixed
        // https://github.com/rofrischmann/inline-style-prefixer/issues/159
        case 'backgroundClip': {
          if (value === 'text') {
            resolvedStyle.backgroundClip = value;
            resolvedStyle.WebkitBackgroundClip = value;
          }
          break;
        }

        case 'display': {
          resolvedStyle.display = value;
          // A flex container in React Native has these defaults which should be
          // set only if there is no otherwise supplied flex style.
          if (style.display === 'flex' && style.flex == null) {
            if (style.flexShrink == null) {
              resolvedStyle.flexShrink = 0;
            }
            if (style.flexBasis == null) {
              resolvedStyle.flexBasis = 'auto';
            }
          }
          break;
        }

        // The 'flex' property value in React Native must be a positive integer,
        // 0, or -1.
        case 'flex': {
          if (value > 0) {
            resolvedStyle.flexGrow = value;
            resolvedStyle.flexShrink = 1;
            resolvedStyle.flexBasis = '0%';
          } else if (value === 0) {
            resolvedStyle.flexGrow = 0;
            resolvedStyle.flexShrink = 0;
            resolvedStyle.flexBasis = '0%';
          } else if (value === -1) {
            resolvedStyle.flexGrow = 0;
            resolvedStyle.flexShrink = 1;
            resolvedStyle.flexBasis = 'auto';
          }
          break;
        }

        case 'font': {
          resolvedStyle[prop] = value.replace('System', systemFontStack);
          break;
        }

        case 'fontFamily': {
          if (value.indexOf('System') > -1) {
            const stack = value.split(/,\s*/);
            stack[stack.indexOf('System')] = systemFontStack;
            resolvedStyle[prop] = stack.join(', ');
          } else if (value === 'monospace') {
            resolvedStyle[prop] = monospaceFontStack;
          } else {
            resolvedStyle[prop] = value;
          }
          break;
        }

        case 'fontVariant': {
          if (Array.isArray(value) && value.length > 0) {
            resolvedStyle.fontVariant = value.join(' ');
          }
          break;
        }

        case 'textAlignVertical': {
          resolvedStyle.verticalAlign = value === 'center' ? 'middle' : value;
          break;
        }

        case 'textDecorationLine': {
          // use 'text-decoration' for browsers that only support CSS2
          // text-decoration (e.g., IE, Edge)
          resolvedStyle.textDecoration = value;
          break;
        }

        case 'transform':
        case 'transformMatrix': {
          resolveTransform(resolvedStyle, style);
          break;
        }

        default: {
          const longFormProperties = styleShortFormProperties[prop];
          if (longFormProperties) {
            longFormProperties.forEach((longForm, i) => {
              // The value of any longform property in the original styles takes
              // precedence over the shortform's value.
              if (typeof style[longForm] === 'undefined') {
                resolvedStyle[longForm] = value;
              }
            });
          } else {
            resolvedStyle[prop] = Array.isArray(value) ? value.join(', ') : value;
          }
        }
      }
    });

  return resolvedStyle;
};

export default createReactDOMStyle;
