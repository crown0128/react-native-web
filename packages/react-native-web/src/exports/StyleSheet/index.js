/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule StyleSheet
 * @noflow
 */

import flattenStyle from './flattenStyle';
import getHairlineWidth from './getHairlineWidth';
import modality from '../../modules/modality';
import ReactNativePropRegistry from '../../modules/ReactNativePropRegistry';
import styleResolver from './styleResolver';

// initialize focus-ring fix
modality();

// allow component styles to be editable in React Dev Tools
if (process.env.NODE_ENV !== 'production') {
  const { canUseDOM } = require('fbjs/lib/ExecutionEnvironment');
  if (canUseDOM && window.__REACT_DEVTOOLS_GLOBAL_HOOK__) {
    window.__REACT_DEVTOOLS_GLOBAL_HOOK__.resolveRNStyle = flattenStyle;
  }
}

const absoluteFillObject = {
  position: 'absolute',
  left: 0,
  right: 0,
  top: 0,
  bottom: 0
};
const absoluteFill = ReactNativePropRegistry.register(absoluteFillObject);

const StyleSheet = {
  absoluteFill,
  absoluteFillObject,
  compose(style1, style2) {
    if (style1 && style2) {
      return [style1, style2];
    } else {
      return style1 || style2;
    }
  },
  create(styles) {
    const result = {};
    Object.keys(styles).forEach(key => {
      if (process.env.NODE_ENV !== 'production') {
        const StyleSheetValidation = require('./StyleSheetValidation').default;
        StyleSheetValidation.validateStyle(key, styles);
      }
      const id = styles[key] && ReactNativePropRegistry.register(styles[key]);
      result[key] = id;
    });
    return result;
  },
  flatten: flattenStyle,
  getStyleSheets() {
    return styleResolver.getStyleSheets();
  },
  hairlineWidth: getHairlineWidth()
};

export default StyleSheet;
