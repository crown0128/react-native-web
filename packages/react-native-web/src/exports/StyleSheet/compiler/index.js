/**
 * Copyright (c) Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 */

import createReactDOMStyle from './createReactDOMStyle';
import hash from './hash';
import hyphenateStyleName from './hyphenateStyleName';
import normalizeValueWithProperty from './normalizeValueWithProperty';
import prefixStyles from '../../../modules/prefixStyles';
import preprocess from './preprocess';

type Value = Object | Array<any> | string | number;
type Style = { [key: string]: Value };
type Rule = string;
type Rules = Array<Rule>;
type RulesData = [Rules, number];
type CompiledStyle = { $$css: boolean, [key: string]: string };
type CompilerOutput = [CompiledStyle, Array<RulesData>];

const cache = new Map();

const classicGroup = 1;
const atomicGroup = 2.2;
const customGroup = {
  borderColor: 2,
  borderRadius: 2,
  borderStyle: 2,
  borderWidth: 2,
  display: 2,
  flex: 2,
  margin: 2,
  overflow: 2,
  overscrollBehavior: 2,
  padding: 2,
  marginHorizontal: 2.1,
  marginVertical: 2.1,
  paddingHorizontal: 2.1,
  paddingVertical: 2.1
};

/**
 * Compile style to atomic CSS rules.
 */
export function atomic(style: Style): CompilerOutput {
  const compiledStyle = { $$css: true };
  const compiledRules = [];

  Object.keys(style)
    .sort()
    .forEach((key) => {
      const value = style[key];
      if (value != null) {
        const valueString = stringifyValueWithProperty(value, key);
        const cacheKey = key + valueString;
        const cachedResult = cache.get(cacheKey);
        if (cachedResult != null) {
          compiledStyle[key] = cachedResult[0];
          compiledRules.push(cachedResult[1]);
        } else {
          const order = customGroup[key] || atomicGroup;
          const identifier = createIdentifier('r', key, value);
          const rules = createAtomicRules(identifier, key, value);
          const orderedRules = [rules, order];
          compiledStyle[key] = identifier;
          compiledRules.push(orderedRules);
          cache.set(cacheKey, [identifier, orderedRules]);
        }
      }
    });

  return [compiledStyle, compiledRules];
}

/**
 * Compile simple style object to classic CSS rules.
 * No support for 'placeholderTextColor', 'scrollbarWidth', or 'pointerEvents'.
 */
export function classic(style: Style, name: string): CompilerOutput {
  const compiledStyle = { $$css: true };
  const compiledRules = [];

  const { animationKeyframes, ...rest } = style;
  const identifier = createIdentifier('css', name, style);
  const selector = `.${identifier}`;
  let animationName;
  if (animationKeyframes != null) {
    const [animationNames, keyframesRules] = processKeyframesValue(animationKeyframes);
    animationName = animationNames.join(',');
    compiledRules.push(...keyframesRules);
  }
  const block = createDeclarationBlock({ ...rest, animationName });
  compiledRules.push(`${selector}${block}`);

  compiledStyle[identifier] = identifier;
  return [compiledStyle, [[compiledRules, classicGroup]]];
}

/**
 * Compile simple style object to inline DOM styles.
 * No support for 'animationKeyframes', 'placeholderTextColor', 'scrollbarWidth', or 'pointerEvents'.
 */
export function inline(style: Style): { [key: string]: mixed } {
  return createReactDOMStyle(style, true);
}

export { preprocess };

/**
 * Create a value string that normalizes different input values with a common
 * output.
 */
export function stringifyValueWithProperty(value: Value, property: ?string): string {
  // e.g., 0 => '0px', 'black' => 'rgba(0,0,0,1)'
  const normalizedValue = normalizeValueWithProperty(value, property);
  return typeof normalizedValue !== 'string'
    ? JSON.stringify(normalizedValue || '')
    : normalizedValue;
}

/**
 * Create the Atomic CSS rules needed for a given StyleSheet rule.
 * Translates StyleSheet declarations to CSS.
 */
function createAtomicRules(identifier: string, property, value): Rules {
  const rules = [];
  const selector = `.${identifier}`;

  // Handle non-standard properties and object values that require multiple
  // CSS rules to be created.
  switch (property) {
    case 'animationKeyframes': {
      const [animationNames, keyframesRules] = processKeyframesValue(value);
      const block = createDeclarationBlock({ animationName: animationNames.join(',') });
      rules.push(`${selector}${block}`, ...keyframesRules);
      break;
    }

    // Equivalent to using '::placeholder'
    case 'placeholderTextColor': {
      const block = createDeclarationBlock({ color: value, opacity: 1 });
      rules.push(
        `${selector}::-webkit-input-placeholder${block}`,
        `${selector}::-moz-placeholder${block}`,
        `${selector}:-ms-input-placeholder${block}`,
        `${selector}::placeholder${block}`
      );
      break;
    }

    // Polyfill for additional 'pointer-events' values
    // See d13f78622b233a0afc0c7a200c0a0792c8ca9e58
    case 'pointerEvents': {
      let finalValue = value;
      if (value === 'auto' || value === 'box-only') {
        finalValue = 'auto!important';
        if (value === 'box-only') {
          const block = createDeclarationBlock({ pointerEvents: 'none' });
          rules.push(`${selector}>*${block}`);
        }
      } else if (value === 'none' || value === 'box-none') {
        finalValue = 'none!important';
        if (value === 'box-none') {
          const block = createDeclarationBlock({ pointerEvents: 'auto' });
          rules.push(`${selector}>*${block}`);
        }
      }
      const block = createDeclarationBlock({ pointerEvents: finalValue });
      rules.push(`${selector}${block}`);
      break;
    }

    // Polyfill for draft spec
    // https://drafts.csswg.org/css-scrollbars-1/
    case 'scrollbarWidth': {
      if (value === 'none') {
        rules.push(`${selector}::-webkit-scrollbar{display:none}`);
      }
      const block = createDeclarationBlock({ scrollbarWidth: value });
      rules.push(`${selector}${block}`);
      break;
    }

    default: {
      const block = createDeclarationBlock({ [property]: value });
      rules.push(`${selector}${block}`);
      break;
    }
  }

  return rules;
}

/**
 * Creates a CSS declaration block from a StyleSheet object.
 */
function createDeclarationBlock(style: Style): string {
  const domStyle = prefixStyles(createReactDOMStyle(style));
  const declarationsString = Object.keys(domStyle)
    .map((property) => {
      const value = domStyle[property];
      const prop = hyphenateStyleName(property);
      // The prefixer may return an array of values:
      // { display: [ '-webkit-flex', 'flex' ] }
      // to represent "fallback" declarations
      // { display: -webkit-flex; display: flex; }
      if (Array.isArray(value)) {
        return value.map((v) => `${prop}:${v}`).join(';');
      } else {
        return `${prop}:${value}`;
      }
    })
    // Once properties are hyphenated, this will put the vendor
    // prefixed and short-form properties first in the list.
    .sort()
    .join(';');

  return `{${declarationsString};}`;
}

/**
 * An identifier is associated with a unique set of styles.
 */
function createIdentifier(prefix: string, name: string, value: Value): string {
  const hashedString = hash(name + stringifyValueWithProperty(value, name));
  return process.env.NODE_ENV !== 'production'
    ? `${prefix}-${name}-${hashedString}`
    : `${prefix}-${hashedString}`;
}

/**
 * Create individual CSS keyframes rules.
 */
function createKeyframes(keyframes) {
  const prefixes = ['-webkit-', ''];
  const identifier = createIdentifier('r', 'animation', keyframes);

  const steps =
    '{' +
    Object.keys(keyframes)
      .map((stepName) => {
        const rule = keyframes[stepName];
        const block = createDeclarationBlock(rule);
        return `${stepName}${block}`;
      })
      .join('') +
    '}';

  const rules = prefixes.map((prefix) => {
    return `@${prefix}keyframes ${identifier}${steps}`;
  });
  return [identifier, rules];
}

/**
 * Create CSS keyframes rules and names from a StyleSheet keyframes object.
 */
function processKeyframesValue(keyframesValue) {
  if (typeof keyframesValue === 'number') {
    throw new Error(`Invalid CSS keyframes type: ${typeof keyframesValue}`);
  }

  const animationNames = [];
  const rules = [];
  const value = Array.isArray(keyframesValue) ? keyframesValue : [keyframesValue];

  value.forEach((keyframes) => {
    if (typeof keyframes === 'string') {
      // Support external animation libraries (identifiers only)
      animationNames.push(keyframes);
    } else {
      // Create rules for each of the keyframes
      const [identifier, keyframesRules] = createKeyframes(keyframes);
      animationNames.push(identifier);
      rules.push(...keyframesRules);
    }
  });

  return [animationNames, rules];
}
