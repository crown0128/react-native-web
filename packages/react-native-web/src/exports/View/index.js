/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import type { ViewProps } from './types';

import createElement from '../createElement';
import css from '../StyleSheet/css';
import setAndForwardRef from '../../modules/setAndForwardRef';
import useElementLayout from '../../hooks/useElementLayout';
import usePlatformMethods from '../../hooks/usePlatformMethods';
import StyleSheet from '../StyleSheet';
import TextAncestorContext from '../Text/TextAncestorContext';
import React, { forwardRef, useContext, useRef } from 'react';

export type { ViewProps };

function createHitSlopElement(hitSlop) {
  const hitSlopStyle = {};
  for (const prop in hitSlop) {
    if (hitSlop.hasOwnProperty(prop)) {
      const value = hitSlop[prop];
      hitSlopStyle[prop] = value > 0 ? -1 * value : 0;
    }
  }
  return createElement('span', {
    classList: [classes.hitSlop],
    style: hitSlopStyle
  });
}

const View = forwardRef<ViewProps, *>((props, ref) => {
  const {
    accessibilityLabel,
    accessibilityLiveRegion,
    accessibilityRelationship,
    accessibilityRole,
    accessibilityState,
    accessibilityValue,
    forwardedRef,
    hitSlop,
    importantForAccessibility,
    nativeID,
    onBlur,
    onContextMenu,
    onFocus,
    onLayout,
    onMoveShouldSetResponder,
    onMoveShouldSetResponderCapture,
    onResponderEnd,
    onResponderGrant,
    onResponderMove,
    onResponderReject,
    onResponderRelease,
    onResponderStart,
    onResponderTerminate,
    onResponderTerminationRequest,
    onScrollShouldSetResponder,
    onScrollShouldSetResponderCapture,
    onSelectionChangeShouldSetResponder,
    onSelectionChangeShouldSetResponderCapture,
    onStartShouldSetResponder,
    onStartShouldSetResponderCapture,
    pointerEvents,
    testID,
    // unstable
    onClick,
    onClickCapture,
    onScroll,
    onWheel,
    onKeyDown,
    onKeyPress,
    onKeyUp,
    onMouseDown,
    onMouseEnter,
    onMouseLeave,
    onMouseMove,
    onMouseOver,
    onMouseOut,
    onMouseUp,
    onTouchCancel,
    onTouchCancelCapture,
    onTouchEnd,
    onTouchEndCapture,
    onTouchMove,
    onTouchMoveCapture,
    onTouchStart,
    onTouchStartCapture,
    href,
    itemID,
    itemRef,
    itemProp,
    itemScope,
    itemType,
    rel,
    target,
    unstable_ariaSet,
    unstable_dataSet
  } = props;

  if (process.env.NODE_ENV !== 'production') {
    React.Children.toArray(props.children).forEach(item => {
      if (typeof item === 'string') {
        console.error(`Unexpected text node: ${item}. A text node cannot be a child of a <View>.`);
      }
    });
  }

  const hasTextAncestor = useContext(TextAncestorContext);
  const hostRef = useRef(null);
  const setRef = setAndForwardRef({
    getForwardedRef: () => forwardedRef,
    setLocalRef: c => {
      hostRef.current = c;
    }
  });

  const children = hitSlop
    ? React.Children.toArray([createHitSlopElement(hitSlop), props.children])
    : props.children;
  const classList = [classes.view];
  const style = StyleSheet.compose(
    hasTextAncestor && styles.inline,
    props.style
  );

  useElementLayout(hostRef, onLayout);
  usePlatformMethods(hostRef, ref, classList, style);

  return createElement('div', {
    accessibilityLabel,
    accessibilityLiveRegion,
    accessibilityRelationship,
    accessibilityRole,
    accessibilityState,
    accessibilityValue,
    children,
    classList,
    importantForAccessibility,
    nativeID,
    onBlur,
    onContextMenu,
    onFocus,
    onMoveShouldSetResponder,
    onMoveShouldSetResponderCapture,
    onResponderEnd,
    onResponderGrant,
    onResponderMove,
    onResponderReject,
    onResponderRelease,
    onResponderStart,
    onResponderTerminate,
    onResponderTerminationRequest,
    onScrollShouldSetResponder,
    onScrollShouldSetResponderCapture,
    onSelectionChangeShouldSetResponder,
    onSelectionChangeShouldSetResponderCapture,
    onStartShouldSetResponder,
    onStartShouldSetResponderCapture,
    pointerEvents,
    ref: setRef,
    style,
    testID,
    // unstable
    onClick,
    onClickCapture,
    onScroll,
    onWheel,
    onKeyDown,
    onKeyPress,
    onKeyUp,
    onMouseDown,
    onMouseEnter,
    onMouseLeave,
    onMouseMove,
    onMouseOver,
    onMouseOut,
    onMouseUp,
    onTouchCancel,
    onTouchCancelCapture,
    onTouchEnd,
    onTouchEndCapture,
    onTouchMove,
    onTouchMoveCapture,
    onTouchStart,
    onTouchStartCapture,
    href,
    itemID,
    itemRef,
    itemProp,
    itemScope,
    itemType,
    rel,
    target,
    unstable_ariaSet,
    unstable_dataSet
  });
});

View.displayName = 'View';

const classes = css.create({
  view: {
    alignItems: 'stretch',
    border: '0 solid black',
    boxSizing: 'border-box',
    display: 'flex',
    flexBasis: 'auto',
    flexDirection: 'column',
    flexShrink: 0,
    margin: 0,
    minHeight: 0,
    minWidth: 0,
    padding: 0,
    position: 'relative',
    zIndex: 0
  },
  // this zIndex-ordering positions the hitSlop above the View but behind
  // its children
  hitSlop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex: -1
  }
});

const styles = StyleSheet.create({
  inline: {
    display: 'inline-flex'
  }
});

export default View;
