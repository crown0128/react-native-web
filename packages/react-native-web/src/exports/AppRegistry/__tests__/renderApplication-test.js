/* eslint-env jasmine, jest */

import ExecutionEnvironment from 'fbjs/lib/ExecutionEnvironment';
import { getApplication } from '../renderApplication';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { render } from 'enzyme';
import StyleSheet from '../../StyleSheet';
import View from '../../View';

const RootComponent = () => <div />;

describe('AppRegistry/renderApplication', () => {
  describe('getApplication', () => {
    const canUseDOM = ExecutionEnvironment.canUseDOM;

    beforeEach(() => {
      ExecutionEnvironment.canUseDOM = false;
    });

    afterEach(() => {
      ExecutionEnvironment.canUseDOM = canUseDOM;
    });

    test('returns "element" and "getStyleElement"', () => {
      const { element, getStyleElement } = getApplication(RootComponent, {});
      expect(element).toMatchSnapshot();
      expect(ReactDOMServer.renderToStaticMarkup(getStyleElement())).toMatchSnapshot();
    });

    test('"getStyleElement" produces styles that are a function of rendering "element"', () => {
      const getTextContent = getStyleElement =>
        getStyleElement().props.dangerouslySetInnerHTML.__html;

      // First "RootComponent" render
      let app = getApplication(RootComponent, {});
      render(app.element);
      const first = getTextContent(app.getStyleElement);

      // Next render is a different tree; the style sheet should be different
      const styles = StyleSheet.create({ root: { borderWidth: 1234, backgroundColor: 'purple' } });
      app = getApplication(() => <View style={styles.root} />, {});
      render(app.element);
      const second = getTextContent(app.getStyleElement);

      const diff = second.split(first)[1];

      expect(first).toMatchSnapshot('CSS for an unstyled app');
      expect(diff).toMatchSnapshot('Additional CSS for styled app');
      expect(first).not.toEqual(second);

      // Final render is once again "RootComponent"; the style sheet should not
      // be polluted by earlier rendering of a different tree
      app = getApplication(RootComponent, {});
      render(app.element);
      const third = getTextContent(app.getStyleElement);

      expect(first).toEqual(third);
    });
  });
});
