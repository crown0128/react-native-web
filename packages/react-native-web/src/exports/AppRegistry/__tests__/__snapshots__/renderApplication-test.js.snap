// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apis/AppRegistry/renderApplication getApplication 1`] = `
<AppContainer
  rootTag={Object {}}
>
  <RootComponent />
</AppContainer>
`;

exports[`apis/AppRegistry/renderApplication getApplication 2`] = `
"<style id=\\"react-native-stylesheet\\">@media all{
html{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0);}
body{margin:0;}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0;}
input::-webkit-inner-spin-button,input::-webkit-outer-spin-button,input::-webkit-search-cancel-button,input::-webkit-search-decoration,input::-webkit-search-results-button,input::-webkit-search-results-decoration{display:none;}
}
@keyframes rn-ActivityIndicator-animation{0%{-webkit-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);transform:rotate(360deg);}}
@keyframes rn-ProgressBar-animation{0%{-webkit-transform:translateX(-100%);transform:translateX(-100%);}100%{-webkit-transform:translateX(400%);transform:translateX(400%);}}</style>"
`;
